packages/@glimmer/reference/lib/reference.ts

```diff
@@ -1,95 +1,273 @@
+  );
+  ref.debugLabel = inner.debugLabel;
+  ref[REFERENCE] = ReferenceType.Invokable;
+
+  return ref;
+}
+
+export function isConstRef(_ref: Reference) {
+  let ref = _ref as ReferenceImpl;
 
-export class ReferenceCache<T> {
-  private reference: Reference<T>;
-  private lastValue: T;
+  return ref.tag === CONSTANT_TAG;
+}
+
+export function isUpdatableRef(_ref: Reference) {
+  let ref = _ref as ReferenceImpl;
+
+  return ref.update !== null;
+}
+
+export function valueForRef<T>(_ref: Reference<T>): T {
+  let ref = _ref as ReferenceImpl<T>;
+
+  let { tag } = ref;
+
+  if (tag === CONSTANT_TAG) {
+    return ref.lastValue as T;
+  }
+
+  let { lastRevision } = ref;
+  let lastValue;
+
+  if (tag === null || !validateTag(tag, lastRevision)) {
+    let { compute } = ref;
+
+    beginTrackFrame(DEBUG && ref.debugLabel);
+    lastValue = ref.lastValue = compute!();
+    tag = ref.tag = endTrackFrame();
+    ref.lastRevision = valueForTag(tag);
+  } else {
+    lastValue = ref.lastValue;
+  }
+
+  consumeTag(tag);
+
+  return lastValue as T;
+}
+
+export function updateRef(_ref: Reference, value: unknown) {
+  let ref = _ref as ReferenceImpl;
+
+  let update = expect(ref.update, 'called update on a non-updatable reference');
+
+  update(value);
+}
 
-  constructor(reference: Reference<T>) {
-    this.reference = reference;
-    this.lastValue = reference.value();
+export function childRefFor(_parentRef: Reference, path: string): Reference {
+  let parentRef = _parentRef as ReferenceImpl;
+
+  let type = parentRef[REFERENCE];
+
+  let children = parentRef.children;
+  let child: Reference;
+
+  if (children === null) {
+    children = parentRef.children = new Map();
+  } else {
+    child = children.get(path)!;
+
+    if (child !== undefined) {
+      if (DEBUG) {
+        // TODO: Make this return a new reference for debug tracing
+        return child;
+      } else {
+        return child;
+      }
+    }
   }
 
-  revalidate(): Validation<T> {
-    let { lastValue } = this;
-    let currentValue = this.reference.value();
+  if (type === ReferenceType.Unbound) {
+    let parent = valueForRef(parentRef);
+
+    if (isDict(parent)) {
+      child = createUnboundRef((parent as Record<string, unknown>)[path]);
+
+      if (DEBUG) {
+        child.debugLabel = `${parentRef.debugLabel}.${path}`;
+      }
+    } else {
+      child = UNDEFINED_REFERENCE;
+    }
+  } else {
+    child = createComputeRef(
+      () => {
+        let parent = valueForRef(parentRef);
+
+        if (isDict(parent)) {
+          return getProp(parent, path);
+        }
+      },
+      val => {
+        let parent = valueForRef(parentRef);
 
-    if (currentValue === lastValue) return NOT_MODIFIED;
-    this.lastValue = currentValue;
+        if (isDict(parent)) {
+          return setProp(parent, path, val);
+        }
+      }
+    );
 
-    return currentValue;
+    if (DEBUG) {
+      child.debugLabel = `${parentRef.debugLabel}.${path}`;
```

Comment:
This should probably confirm that `parentRef.debugLabel` is not `false` before chaining

Reply:
it will only every be `false` in production, is the idea, so it should always be a string here. I guess we could assert that here